Add Heuristic CTE Materialization Strategy

jaystarshot opened the pull request:
1. This PR adds support Heuristic CTE materialization.
We have rolled this in one of our production cluster and confirmed that this is stable. 
Our current testing has shown **upto** 50% cpu gain and > 50% less hdfs read for affected queries.
Its heuristic so there may be cases where non-ideal plans are generated.

2. Changed cteName member variable to cteId for cte plan nodes.

3. Also fixed CteInformation to include cteId so that correct references are incremented which crucial for cte materialization.


In heuristic strategy, the relational planner adds cteReferences everywhere and the LogicalCteOptimizer decides which references to implement


Fixes #21637 

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Add Heuristic CTE Materialization strategy which auto materialized expensive ctes. This is configurable by setting ``cte_materialization_strategy`` to ``HEURISTIC`` or ``HEURISTIC_COMPLEX_QUERIES_ONLY``. (default ``NONE``)

jaystarshot commented:
@feilong-liu  @mlyublena Please review when you get time, thanks!

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -357,7 +357,9 @@ public boolean isAdoptingMergedPreference()
         ALL, // Materialize all CTES
-        NONE // Materialize no ctes
+        NONE, // Materialize no CTES
+        HEURISTIC,
```
can you add comments for the new values

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -357,7 +357,9 @@ public boolean isAdoptingMergedPreference()
-        NONE // Materialize no ctes
+        NONE, // Materialize no CTES
+        HEURISTIC,
+        HEURISTIC_STRICT
```
I wonder if there is a better name than strict: perhaps HEURISTIC_COMPLEX_QUERIES_ONLY? we already have "complex query" reference in another session property (RESTRICT_HISTORY_BASED_OPTIMIZATION_TO_COMPLEX_QUERY)

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java:
```diff
@@ -132,34 +136,65 @@ public boolean isPlanRewritten()
+            HashMap<String, CTEInformation> cteInformationMap = session.getCteInformationCollector().getCteInformationMap();
+            CTEInformation cteInfo = cteInformationMap.get(node.getCteId());
+            switch (getCteMaterializationStrategy(session)) {
+                case HEURISTIC_STRICT:
```
do we want to also disable materialization of VALUES nodes? so only proceed if the query is a proper join/aggregation on top of table scans

mlyublena commented:
LGTM modulo some nits

jaystarshot commented:
@mlyublena  Thanks for your review, I have 
1. updated the PR to include the values check
2. Added tests in case to assert no materialization in heuristic cases
3. Backported a small fix from our repo in case there is no materialization in heuristic cases (cte references were not removed)

jaystarshot commented:
Rebased

mlyublena commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -1082,6 +1083,11 @@ public SystemSessionProperties(
                         false),
+                integerProperty(
+                        CTE_HEURISTIC_REPLICATION_THRESHOLD,
+                        "Used with CTE Materialization Strategy = Heuristic. CTES are only materialized if they are used greater than or equal to this number",
```
nit: perhaps change comment to "if they are used more than this number of times"

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java:
```diff
@@ -132,34 +139,68 @@ public boolean isPlanRewritten()
+            CTEInformation cteInfo = cteInformationMap.get(node.getCteId());
+            switch (getCteMaterializationStrategy(session)) {
+                case HEURISTIC_COMPLEX_QUERIES_ONLY:
+                    if (!PlanNodeSearcher.searchFrom(node).where(planNode ->
```
perhans refactor this expression into a function containsNodeType(node, NodeTypesList)

also: should we recursively handle CTEConsumers here? Right now we allow this if we see at least one CTE consumer as data source, but the cte producer for that one may be a trivial one (like a table scan)?

jaystarshot replied:
Actually heuristic strategy flow for dependent CTEs is also not exactly accurate right now (eg: if a inner cte having a join/agg is materialized, the outer cte using this inner one might not be complex). This would need a completly different implementation which we would need to address in the cost based stategy. I was wondering if we could keep heuristic strategy as simple as possible

jaystarshot replied:
PlanNodeSearcher.searchFrom(node) searches the node with CTEReferences, those are not rewritten as CteConsumers when this search happens

jaystarshot replied:
If we visit the children of this node first and convert them to cte producers that might fix the problem. 
Also` Right now we allow this if we see at least one CTE consumer as data source, but the cte producer for that one may be a trivial one (like a values node)?` will never happen right? Since the heuristic strategy will ensures that the inner cte will never be converted to a cte producer.

jaystarshot replied:
I fixed this test case in the refactor

mlyublena commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java:
```diff
@@ -132,34 +139,68 @@ public boolean isPlanRewritten()
-        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        private final Session session;
+
+        private static final List<Class<? extends PlanNode>> PRECOMPUTE_PLAN_NODES = ImmutableList.of(JoinNode.class, SemiJoinNode.class, AggregationNode.class);
```
should we call these two COMPLEX_PLAN_NODES and COMPLEX_DATA_SOURCE_NODES or something like that to better correlate with the session property name?

jaystarshot commented:
@mlyublena @pranjalssh Thank you for the review. However I found a test case added [here](https://github.com/prestodb/presto/pull/21720/files#diff-f1b72dc5b02cc72edb905847f0c8fe2b5599a929349f00ce8c902724a9469138R315) wherein, the CTE would be materialized even if it wasn't complex if the inner CTE was complex,
hence I changed the flow a bit to accommodate this.
1. Now, we first determine if a CTE is complex and then perform the rewrite.
2. Also cleaned up some session property handling, along with some renaming in the classes for clarity.
Please re-review!
cc: @tdcmeehan

jaystarshot commented:
@pranjalssh Need a review again due to the comment change
